src/main/scala/scala/async/internal/TransformUtils.scala

```diff
@@ -362,4 +386,102 @@ private[async] trait TransformUtils {
     else withAnnotation(tp, Annotation(UncheckedBoundsClass.asType.toType, Nil, ListMap()))
   }
   // =====================================
+
+  /**
+   * Efficiently decorate each subtree within `t` with the result of `t exists isAwait`,
+   * and return a function that can be used on derived trees to efficiently test the
+   * same condition.
+   *
+   * If the derived tree contains synthetic wrapper trees, these will be recursed into
+   * in search of a sub tree that was decorated with the cached answer.
+   */
+  final def containsAwaitCached(t: Tree): Tree => Boolean = {
+    val symtab = c.universe.asInstanceOf[scala.reflect.internal.SymbolTable]
+    def attachContainsAwait(t: Tree): Unit = {
+      val t1 = t.asInstanceOf[symtab.Tree]
+      t1.updateAttachment(ContainsAwait)
+      t1.removeAttachment[NoAwait.type]
+    }
+    def attachNoAwait(t: Tree): Unit = {
+      val t1 = t.asInstanceOf[symtab.Tree]
+      t1.updateAttachment(NoAwait)
+    }
+    object markContainsAwaitTraverser extends Traverser {
+      var stack: List[Tree] = Nil
+
+      override def traverse(tree: Tree): Unit = {
+        stack ::= tree
+        try {
+          if (isAwait(tree))
+            stack.foreach(attachContainsAwait)
+          else
+            attachNoAwait(tree)
```

Comment:
Could indeed trivially skip `Ident`, also `TypeTree`.

`Select`-s might contain an `await` in the qualifier. I guess we could omit the attachment and always traverse it as there is only one child, so the caching wouldn't pay its way.

Another thing I considered is stripping the attachments at the end of the macro.